### PR TITLE
Implement reparameterized gradient for Gamma sampler

### LINKED ISSUE #3813: Implement random_gamma() sampler (with gradients)
Many probability distributions require sampling from the [Gamma distribution](https://en.wikipedia.org/wiki/Gamma_distribution), including: `Gamma`, `Beta`, and `Dirichlet`.

Since Gamma samplers have complex control flow (for rejection sampling) and are seldom a bottleneck in probabilistic algorithms, it should suffice to implement a CPU-only implementation at first. What is more important than a CUDA implementation is a reparameterized sampler so that stochastic gradients can be propagated through the sampler (see [paper](http://proceedings.mlr.press/v54/naesseth17a/naesseth17a.pdf) and [reference implementation](https://github.com/blei-lab/ars-reparameterization) by @naesseth).

## Tasks
- [x] #3841 CPU implementation of basic `random_gamma(requires_grad=False)`
- [ ] #3978 Support reparameterized `random_gamma(requires_grad=True)`

## Map of modifications
- `aten/src/TH/THRandom.c/h` random single numbers
- `aten/src/TH/generic/THTensorRandom.c/h` random tensors
- `aten/src/ATen/Declarations.cwrap` bindings for ATen
- `torch/csrc/generic/methods/TensorRandom.cwrap` bindings for torch.Tensor
- `torch/autograd/variable.py` - Variable
- `torch/distributions.py` - Distributions

### PR DESCRIPTION
**Pull request by fritzo**
Closes #3813

This implements reparameterized gradient for `distributions.Gamma`. The gradient is implemented by directly approximating the reparameterized gradient function `dx/dalpha` following [Knowles (2015)](https://arxiv.org/pdf/1509.01631.pdf). The approximation is accurate to within 0.5% relative error for a wide range of alphas.

## Derivation

First consider the `beta` variable. If `x ~ Gamma(alpha, beta)` then `x / beta ~ Gamma(alpha, 1)`. Since division is already implemented in PyTorch, we can thus reduce our problem to computing a reparameterized gradient of a standard gamma `x ~ Gamma(alpha) = Gamma(alpha, 1)` wrt `alpha`.

This PR implements a function `standard_gamma_grad(x, alpha)` that directly approximates the reparameterized gradient defined (for any continuous univariate distribution) as
```
                d/dalpha cdf(x; alpha)     d/dalpha cdf(x; alpha)
dx / dalpha = - ---------------------- = - ----------------------
                  d/dx cdf(x; alpha)           pdf(x; alpha)
```
This definition is used in the unit tests in `tests/test_distributions.py`, which compute `d/dalpha cdf(x;alpha)` via finite difference of the `scipy.stats.gamma.cdf()` function.

The approximation is split into three regions:
- For small `x` we use a power series approximation of `cdf(x, alpha)`.
  Until `digamma()` is implemented in PyTorch, we use a finite difference of `lgamma()`.
- For large `alpha` we use the approximation
  ```
  standard_gamma_grad(x, alpha) = sqrt(x/alpha)
  ```
- For intermediate x,alpha we use a rational function approximation
   ```
   standard_gamma_grad(x, alpha) = exp(PQ(log(x / alpha), log(alpha)))
   ```
   where `PQ(u,v)` is a rational function of order 2 in u and 3 in v.

For complete derivation, see this [Jupyter Notebook](https://github.com/fritzo/notebooks/blob/master/gamma-reparameterized.ipynb).

### CODE REVIEW

**Comment by fritzo** [comment]
CC: @apaszke and @martinjankowiak who has reviewed some of the math.

File: aten/src/TH/THRandom.h [hunk 1]
```diff
@@ -71,6 +71,11 @@ TH_API double THRandom_exponential(THGenerator *_generator, double lambda);
 */
 TH_API double THRandom_standard_gamma(THGenerator *_generator, double alpha);
 
+/** Computes a reparameterized gradient of a sample from a standard Gamma
+    distribution wrt the shape parameter alpha.
+*/
+TH_API double THRandom_standard_gamma_grad(double x, double alpha);
```
Comment (apaszke) [inline aten/src/TH/THRandom.h]: I think it would be cleaner to just put the body of this function in the apply macro in THTensorRandom. All other functions in this file do actual sampling, while this is deterministic, and unlikely to be used outside of this single method (correct me if I’m wrong).

File: torch/distributions.py [hunk 2]
```diff
@@ -198,10 +198,27 @@ def log_prob(self, value):
...
+        return grad_output * grad
+
+
 def _standard_gamma(alpha):
     if not isinstance(alpha, Variable):
         return torch._C._standard_gamma(alpha)
-    return Variable(torch._C._standard_gamma(alpha.data))
+    if not alpha.requires_grad:
+        return Variable(torch._C._standard_gamma(alpha.data))
+    return _StandardGamma.apply(alpha)
```
Comment (apaszke) [inline torch/distributions.py]: You should be able to always apply `_StandardGamma` on Variables of all kinds

File: torch/autograd/variable.py [hunk 3]
```diff
@@ -304,6 +304,11 @@ def multinomial(self, num_samples=1, replacement=False):
     def bernoulli(self):
         return Variable(torch.bernoulli(self.data))
 
+    def standard_gamma(self, grad=None):
+        if grad is None:
+            return Variable(torch.standard_gamma(self.data))
+        return Variable(torch.standard_gamma(self.data, grad), requires_grad=self.requires_grad)
```
Comment (apaszke) [inline torch/autograd/variable.py]: Isn’t this one hidden? I think these are some leftovers from previous diffs

**Review by apaszke — COMMENTED** [review]
Looks good for the most part, but changes in variable.py need to be reverted. Haven’t reviewed the math.

File: aten/src/TH/THRandom.h (hunk 1 above)
Comment (fritzo) [reply aten/src/TH/THRandom.h]: Thanks for the suggestion. I've gone one step further and moved this to `THTensorMath.c` so it can use `THTensor(digamma_one)` internally once https://github.com/pytorch/pytorch/pull/3955 is merged. This seems like a better location than `THTensorRandom` because as you observe there is no random generation going on: this is merely a useful special function.

**Comment by fritzo** [comment]
Thanks for reviewing @apaszke and sorry for the slow response. All comments now addressed.

**Review by apaszke — APPROVED** [review]

**Comment by naesseth** [comment]
If you are OK with a (small) bias I believe using the shape augmentation trick in my paper and just ignoring the score function term will be much more accurate and efficient.

https://arxiv.org/abs/1610.05683
